checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -1,146 +1,763 @@
+    if (methodElt != null) {
+      List<? extends VariableTree> paramsOfCurrentMethod = methodTree.getParameters();
+      addOwningParamsFromMethodCall(obligations, invocation, paramsOfCurrentMethod);
+      addOwningReceiverFromMethodCall(obligations, invocation, paramsOfCurrentMethod);
+      checkIndirectCalls(obligations, invocation, paramsOfCurrentMethod);
+    }
+  }
 
-      // This assumes that any MustCall annotation has at most one element.
+  /**
+   * Checks if a MustCall obligation is satisfied via the given method call. A MustCall obligation
+   * of an element is satisfied if the called-methods set contains the target of its must-call
+   * obligation.
+   *
+   * @param invocation the method invocation node being checked for satisfaction of the MustCall
+   *     obligation
+   * @param varElt the variable annotated with the MustCall annotation
+   * @param target the target of the MustCall obligation
+   * @return {@code true} if the MustCall obligation is satisfied
+   */
+  private boolean mustCallObligationSatisfied(
+      MethodInvocationNode invocation, Element varElt, JavaExpression target) {
+
+    List<String> mustCallValues = typeFactory.getMustCallValue(varElt);
+    if (mustCallValues.size() != 1) {
       // TODO: generalize this to MustCall annotations with more than one element.
-      if (mustCallValues.size() == 1
-          && mustCallValues.contains(method.getSimpleName().toString())) {
-        owningFields.add((VariableElement) receiverEl);
+      return false;
+    }
+
+    AccumulationStore cmStoreAfter = typeFactory.getStoreAfter(invocation);
+    @Nullable AccumulationValue cmValue =
+        cmStoreAfter == null ? null : cmStoreAfter.getValue(target);
+    AnnotationMirror cmAnno = null;
+
+    if (cmValue != null) { // When store contains the lhs
+      Set<String> accumulatedValues = cmValue.getAccumulatedValues();
+      if (accumulatedValues != null) { // type variable or wildcard type
+        cmAnno = typeFactory.createCalledMethods(accumulatedValues.toArray(new String[0]));
+      } else {
+        for (AnnotationMirror anno : cmValue.getAnnotations()) {
+          if (AnnotationUtils.areSameByName(
+              anno, "org.checkerframework.checker.calledmethods.qual.CalledMethods")) {
+            cmAnno = anno;
+          }
+        }
       }
     }
+
+    if (cmAnno == null) {
+      cmAnno = typeFactory.top;
+    }
+
+    return mcca.calledMethodsSatisfyMustCall(mustCallValues, cmAnno);
   }
 
   /**
-   * Updates {@code worklist} with the next block along all paths to the regular exit point. If the
-   * next block is a regular exit point, adds an {@literal @}Owning annotation for fields in {@link
-   * #owningFields}.
+   * Adds all non-exceptionsal successors to {@code worklist}. If a successor is a non-exceptional
+   * exit point, adds an {@literal @}Owning annotation for fields in {@link #releasedFields}.
    *
+   * @param obligations the set of obligations to update
```

Comment:
Throughout the code, do "add an `@Owning` annotation" and "update the set of obligations" mean the same thing?  If so, pick the best description and consistently stick with that wording.  If not, clarify the difference, and use each phrase appropriately.  Here they are both used, which is confusing.

Reply:
They are not the same. `obligations` here is the obligations of the current block.